Fixed the Show SQL query action

mickaelandrieu opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | The "Show SQL query" action was wrong, it showed the prepared SQL query instead of the complete one.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Access (again) to the logs page (http://prestashop.local/admin-dev/index.php/configure/advanced/logs/), add a filter to the query and read the SQL query: you should see the value of your filter.

<!-- Click the form's "Preview button" to make sure the table is functional in GitHub. Thank you! -->

![updated](https://user-images.githubusercontent.com/1247388/45171956-1400a500-b204-11e8-8a3a-1bdc4c82eef0.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9324)
<!-- Reviewable:end -->

mickaelandrieu committed:
Fixed the Show SQL query action

mickaelandrieu commented:
One review please @PrestaShop/prestashop-core-dev ? I'd like to see this merged today :+1:

mickaelandrieu committed:
Update tests to add check on query parameters

PierreRambaud commented on tests/Unit/Core/Grid/DataProvider/DoctrineGridDataProviderTest.php:
```diff
@@ -82,7 +82,12 @@ private function createDoctrineQueryBuilderMock()
             ->willReturn($statement);
         $qb->method('getSQL')
-            ->willReturn('SELECT * FROM ps_test');
+            ->willReturn('SELECT * FROM ps_test WHERE id = :id');
```
You must add more tests like
```
'name' => 'Something you\'ll like',
'status' => null,
'values' => [1, 2, 3, 4],
'in_stock' => false,
```
The raw query will have another look :)

PierreRambaud commented on src/Core/Grid/DataProvider/DoctrineGridDataProvider.php:
```diff
@@ -66,7 +67,23 @@ public function getData(SearchCriteriaInterface $searchCriteria)
+        $parameters = $queryBuilder->getParameters();
+
+        foreach ($parameters as $pattern => $value) {
+            $query = str_replace(":$pattern", $value, $query);
```
A function like this one is maybe better (need refactoring, first copy & paste from stackoverflow :P):
```php
public function interpolateQuery($query, $params) {
    $keys = array();
    $values = $params;

    # build a regular expression for each parameter
    foreach ($params as $key => $value) {
        if (is_string($key)) {
            $keys[] = '/:'.$key.'/';
        } else {
            $keys[] = '/[?]/';
        }

        if (is_string($value))
            $values[$key] = "'" . $value . "'";

        if (is_array($value))
            $values[$key] = "'" . implode("','", $value) . "'";

        if (is_null($value))
            $values[$key] = 'NULL';
    }

    $query = preg_replace($keys, $values, $query);

    return $query;
}
```

mickaelandrieu replied:
I don't want to address every use case from the earth until we meet an error, this is how you basically be late with releases and add some dead code in your project.

So show me a real use case when the search can pass an array of data in the search form of logs page Grid and I'll update accordingly, tests comprised. Passing null or `[?]` is impossible by design.

But the day we do it, we will extract this business logic into his own class (and tests)

PierreRambaud replied:
We offer the capability to everyone to add some code with hook or using a custom Filters, so we must be able to parse default values (boolean, string, array, integer at least), it's not every use case, it's the default use case of your function otherwise, it's not returning a raw query :)
More, there is nothing in the design that prevent null, and `[?]` is not a value, just a replacement key in pcre patterns :P

mickaelandrieu replied:
> So show me a real use case when the search can pass an array of data in the search form of logs page Grid and I'll update accordingly, tests comprised.

Remember the label of this pull request: this is not "Make a reusable, tested and complete framework to manage every possible parameter of Doctrine query, even the ones that may not comes from a search form of Grid component" :D

eternoendless replied:
Unless we can guarantee that the `:pattern` pattern is the only one that will be fed to this code, or that the code won't break down if not, then it must be adapted.

mickaelandrieu replied:
Ì can add a constraint to allow only string values, but I don't want to create a "reusable, tested and complete framework to manage every possible parameter of Doctrine query": is it ok for you? This won't break and this won't be supported (for now).

Note that we can't avoid people to construct a bad Doctrine query, if they want to do it it will break the same way the can break a simple SQL query: I'm a great developer, not a wizard ;p

I'm still waiting for a use case where this can happens, as parameters come from the query parameters and so on are strings by design: surprise me 👍

eternoendless replied:
If I understood it right, this code tries to parse prepared statements as MySQL would. The problem happens if another kind of token different from `:pattern` is used, like `?`. It's not magic, it's per spec.

mickaelandrieu replied:
I don't contest the spec, I'm saying I need a concrete use case so I can update my code accordingly ^^ I won't update it if I can't test it: implementing something "just in case" is maybe the biggest part of our technical debt reason right now :+1: 

So, please, break my code :dagger: 

Edit: it's not possible to do it right now as the query builder is never available using hooks in the Grid system workflow.

Quetzacoalt91 replied:
There is no easy solution to get the request sent to MySQL.
We can stick to @mickaelandrieu solution (after rebase), or updating the feature by separating the display of the prepared request and the parameters ready to be applied.

However, I see in the screenshot you apply strings in the request without quotes, and this will break the SQL execution. :smile:

PierreRambaud reviewed (dismissed):
see comments

mickaelandrieu commented:
ping @eternoendless, wdyt? Can we keep the scope of this pull request limited please?

mickaelandrieu commented:
Note this is required to be shipped with `1.7.5.0`.

eternoendless commented:
Why does such a thing exist in the first place?

mickaelandrieu commented:
@eternoendless it's part of legacy... Before the raw SQL was sent from the server to the DOM as sql-query attribute and re-used => https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Resources/views/Admin/Product/CatalogPage/Lists/products_table.html.twig#L198

mickaelandrieu committed:
Fixed coding styles

mickaelandrieu committed:
Abstract the parsing of query from the Grid Data Factory

mickaelandrieu commented:
Hi there, it's now rebased and improved following the proposal of @PierreRambaud.

Also, I've extracted the parsing of the Query from Grid Data Factory and abstracted it behind an interface: this means that if someone finds an issue later because of this we are able to change the implementation use and replace by a better one. 

More, PrestaShop developers are also able to change its [implementation](https://github.com/PrestaShop/PrestaShop/pull/9324/files#diff-97ef2b55b06d437323783fe085e81d33R162) to fit their needs and are not stuck.

So... I guess we're "good enough" here, am I right?

mickaelandrieu committed:
Updated service definitions

mickaelandrieu committed:
Added missing PHPDoc

jolelievre reviewed (dismissed):
is it used to display le last sql query in the back-office? for debug purposes?
anyway that's good to me! 👍

sarjon commented on src/Core/Grid/Query/DoctrineQueryParser.php:
```diff
@@ -0,0 +1,66 @@
+/**
+ * This class offers a DBAL implementation of Query parser.
+ */
+class DoctrineQueryParser implements QueryParserInterface
```
is it not final on purpose?

mickaelandrieu replied:
yes, because we are already aware this could create some issues so people may have the need to update it.

But as it's only 1 function class, I'm wondering if it's worth?

mickaelandrieu committed:
Fixed QueryParserInterface mock

mickaelandrieu committed:
Make DoctrineQueryParser final

PierreRambaud commented:
@jolelievre Not only for debug, also when you're on database request sql page.
@mickaelandrieu I fixed tests, waiting for ci and I will add Waiting for QA.

mickaelandrieu committed:
Added tests and fixed unhandled use case

mickaelandrieu commented:
Sorry for your commit @PierreRambaud, I've found better tests to do: wdyt?

mickaelandrieu commented on src/Core/Grid/Exception/UnsupportedParameterException.php:
```diff
@@ -0,0 +1,33 @@
+
+use Exception;
+
+class UnsupportedParameterException extends Exception implements ExceptionInterface
```
I wanted to add PHPDoc, but what information could we add except "throw an exception in case of unsupported parameter?"

PierreRambaud commented on tests/Unit/Core/Grid/Query/DoctrineQueryParserTest.php:
```diff
@@ -0,0 +1,149 @@
+
+        $expectedQuery = "SELECT tests FROM pierre_rambaud WHERE motivation = 'OK' AND energy = 'none'";
+
+        $this->assertSame($expectedQuery, $this->queryParser->parse($preparedQuery, $queryParameters));
```
You can remove the assertSame and diretly call `->parse`

PierreRambaud commented on tests/Unit/Core/Grid/Query/DoctrineQueryParserTest.php:
```diff
@@ -0,0 +1,149 @@
+
+        $expectedQuery = "SELECT tests FROM pierre_rambaud WHERE motivation IN ('great', 'good', 'ok', 'nok', 'none')";
+
+        $this->assertSame($expectedQuery, $this->queryParser->parse($preparedQuery, $queryParameters));
```
same here

mickaelandrieu commented:
big up @PierreRambaud, you were right to insist on this subject: this is far better now :+1:

mickaelandrieu committed:
Improved DoctrineQueryParser tests

PierreRambaud commented:
Tell me when it's ok to merge?

eternoendless commented on src/Core/Grid/Query/DoctrineQueryParser.php:
```diff
@@ -0,0 +1,127 @@
+
+        $query = preg_replace($keys, $values, $query);
+
+        return $query;
```
There are several issues with this implementation

1. It goes through each `if` even after the type of `$value` has already been found
2. It copies the `$values` needlessly (altering the passed-by-parameter array inside the function requires the array to be copied in memory, then altered)
3. It's needlessly complex
4. It will break if $value cannot be casted to string (for example if it's an object)
5. It uses regex replace needlessly, and keys aren't escaped for special characters

This can be improved as follows:

```php
public function parse($query, array $queryParameters)
{
    $replacements = [];

    foreach ($queryParameters as $key => $value) {
        if (!is_string($key)) {
            throw new UnsupportedParameterException('Only named parameters are supported in prepared queries.');
        }
        $values[':' . $key] = $this->parseValue($value);
    }

    return strtr($valueds, $query);
}

private function parseValue($value)
{
    if (is_string($value)) {
        return $this->parseStringParameter($value);
    }

    if (is_numeric($value)) {
        return $this->parseNumericParameter($value);
    }
    
    if (is_array($value)) {
        return $this->parseArrayParameter($value);
    }

    if (is_bool($value)) {
        return $this->parseBooleanParameter($value);
    }
    
    if ($value === null) {
        return $this->parseNullParameter($value);
    }

    throw new UnsupportedParameterException('Unsupported value type: ' . gettype($value));
}

private function parseStringParameter($value)
{
    // this is wrong but we have no choice
    return = "'" . addslashes($value) . "'";
}

private function parseArrayParameter(array $value)
{
    return "'" . implode("', '", array_map('addslashes', $value)) . "'";
}
```

PierreRambaud commented:
I love what @eternoendless said, could you have a check on it @mickaelandrieu ?

mickaelandrieu committed:
Improved the query parser

mickaelandrieu committed:
Improved the query parser

PierreRambaud commented:
Really nice jog @mickaelandrieu

marionf commented:
@mickaelandrieu 

I have an exception on the log page 

![capture d ecran_291](https://user-images.githubusercontent.com/13449658/45227691-5dfb9080-b2c1-11e8-97ef-ae72d1d57f1f.png)

PierreRambaud commented on src/Core/Grid/Query/DoctrineQueryParser.php:
```diff
@@ -0,0 +1,112 @@
+     */
+    public function parse($query, array $queryParameters)
+    {
+        foreach ($queryParameters as $key => $value) {
```
$values must be initialized :)

mickaelandrieu replied:
it's @eternoendless' fault! :trollface:

mickaelandrieu committed:
Fixed undefined variable

mickaelandrieu commented:
Hello @marionf , it's fixed now :+1:

mickaelandrieu committed:
fixed DoctrineGridDataFactoryTest typo

PierreRambaud commented:
Thanks @mickaelandrieu and everyone for reviews :)
